src/Gql.js

```diff
@@ -0,0 +1,215 @@
+	// we are not using the inbuilt graphql function because it validates
+	// the graphql, which is an expensive operation
+	// return graphql(schema, query, rootValue, contextValue, variableValues);
+	// taken from: https://github.com/graphql/graphql-js/blob/master/src/graphql.js
+
+	// parse
+	let document;
+	try {
+		document = parse(query);
+	}
+	catch (syntaxError) {
+		return Promise.resolve({errors: [syntaxError]});
+	}
+
+	// validate only in developement
+	if (process.env.NODE_ENV !== 'production') {
+		const validationErrors = validate(schemaObj, document);
+		if (validationErrors.length > 0) {
+			return Promise.resolve({errors: validationErrors});
+		}
+	}
+
+	return execute(
+		schemaObj,
+		document,
+		rootValue,
+		contextValue,
+		variableValues,
+	);
+}
+
+function convertObjToGqlArg(obj) {
+	const gqlArg = [];
+	_.forEach(obj, (value, key) => {
+		// eslint-disable-next-line no-use-before-define
+		gqlArg.push(`${key}: ${convertToGqlArg(value)}`);
+	});
+	return `${gqlArg.join(', ')}`;
+}
+
+function convertStrToGqlArg(str) {
+	const [, val] = str.match(ENUM_REGEX) || [];
+	if (val) return val;
+	// const [, prefix, val] = str.match(NO_QUOTES_REGEX) || [];
+	// if (prefix && val) return prefix === VAR_PREFIX ? '$' + val : val;
+
+	return JSON.stringify(str);
+}
+
+function convertToGqlArg(value) {
+	if (value == null) return null;
+
+	if (typeof value === 'number') return String(value);
+	if (typeof value === 'string') return convertStrToGqlArg(value);
+	if (_.isPlainObject(value)) return `{${convertObjToGqlArg(value)}}`;
+
+	return JSON.stringify(value);
+}
+
+class Gql {
+	constructor(opts = {}) {
+		const {schema, pubsub, defaultSchema} = makeSchemaFromConfig(opts);
+
+		this.schema = schema;
+		this.defaultSchema = defaultSchema;
+		this.pubsub = pubsub;
+
+		this.logger = opts.logger || console;
+		this.cache = opts.cache;
+		this.formatError = opts.formatError || formatError;
+	}
+
+	async exec(query, context, {
+		cache: {key: cacheKey, ttl = ONE_DAY} = {},
+		variables = {},
+	} = {}) {
+		if (cacheKey && this.cache) {
+			const cached = await this.cache.get(cacheKey);
+			if (cached) return cached;
+		}
+
+		if (!/^\s*query|mutation|subscription/.test(query) && /^\s*[a-zA-Z0-9]/.test(query)) {
+			query = `query { ${query} }`;
+		}
+
+		const result = await graphql(this.defaultSchema, query, null, context, variables);
+
+		if (!result.errors) {
+			if (cacheKey && this.cache) await this.cache.set(cacheKey, result.data, {ttl});
+			return result.data;
+		}
+
+
+		let fields = {};
+		const errors = result.errors;
+
+		// log & format all errors
+		errors.forEach((error) => {
+			error = this.formatError(error);
+			Object.assign(fields, error.fields);
+		});
+
+		// no user errors sent by server
+		if (!Object.keys(fields).length) {
+			fields = {
+				global: {
+					message: 'Unknown Error',
+					keyword: 'unknown',
+				},
+			};
+		}
+
+		const err = new Error('Error in graphQL api');
```

Comment:
could mention schema in error message, should be available in `opts.schema` or through `getConfig()`